### DISCUSSION TITLE
Extending Budibase

### DISCUSSION THREAD

**Discussion by mjashanks** [Budibase/budibase#890] [Ideas]
So far, we have built an easy to use visual development application - the Budibase Builder.

Next, we want to create a fantastic development experience around the builder, allowing developers to easily extend the platform - with the full power of code.

The potential areas for extension are as follows...
- UI Components. Code your own components (in Svelte), and use them in the builder
- Automations blocks. Write your own backend blocks (in nodejs), and import into the builder.
- External data connectors. Similar to adding a new table, we will be able to add connections to other datasources. e.g. MySQL, GitHub API. We want developer to be able to create their own.
- Middleware. Budibase server is built with Koa. We want to be able to plugin custom Koa middleware.

We want to create an amazing developer experience around these extension points, but what does this look like ?

Since the builder runs locally, we have the option of basing this around the file system - allowing developers to code using their favourite IDE. Is this the way to go ? Should we be aiming to keep the coding experience inside the builder ?

**Comment by gustavopch** [comment]
Please let me use my preferred IDE and my preferred Git provider to write extensions/plugins. You'll surely not be able to provide the same developer experience that my preferred set of tools already provide. Focus on the important stuff that makes Budibase unique, not on trying to compete with VSCode or GitHub. There's so much more you can do than wasting time with that.

**Reply by CallMeLaNN** [reply]
Totally agreed with that. Allowing developers to create and publish extensions/plugins from their own ways as long as they follow the API like npm package naming, the index.js export format and how admin can import the extensions/plugins.

However probably the OP above thinking to have a quick and simple way for admin to extend that, which I also did not recommend that because the possibility might be limited.

Better give developers fully utilize Svelte and NodeJS with given some context like reusing theme, current Svelte state etc.

**Reply by mjashanks** [reply]
Honestly, we are going to get this done. It's been on the list for ages, as you can see. Very important for us going forward - but its is a big piece of work.

**Comment by halfk1ng** [comment]
I would argue that your target persona is not necessary a developer, but a tech savvy individual who understands the constructs of development. Many developers tend to learn a flavor of development, and stick to what's in their wheelhouse. They explore other flavors when their current ones arent doing it for them, or when they have down time. Often, learning something new requires additional overhead rather than just getting it done with what you know.

For someone like myself, taught myself a few front end languages when I was young. I didnt keep up with the evolution of those languages or the syntax. I can write pseudo code for days. My point is that I have enough knowledge to leverage this type of tool without knowing how to code.

I work at a medium-ish enterprise saas company and I've been exposed to a lot of the user experience research. While I cant go into detail for obvious reasons, low-code/no-code environments are most adopted by exactly the overhead which your products remove; the ability to effectively create something without roadblocks, googling for hours, debugging; for a user base that doesnt already do that.

e; words are hard. i fixed words.

e2; please understand I am not trying to make you something that you're not. I like what you are, already. I just want to expose the possibility, or start the discussion, that your product serves a much wider audience than what I'm reading, or interpreting.

**Reply by mjashanks** [reply]
Hey @halfk1ng.

I love your question because it fits perfectly with our world view and vision for Budibase.

We have 2 ideal personas

1. Developer.
2. IT Professional  (aka technical non-developer).

**Developer**
As you correctly point out, there are many flavours of developer. Front end only, backend only, full-stack, developers just out of college, developers who write code for a living but not for a hobby - and many other permutations. 

The Budibase builder needs to serve as the lowest common denominator, between all these types of developer. Thus we have purposely built it to be usable by anyone who is, like you say, "a tech-savvy individual who understands the constructs of development."

Developers are important as a target persona because

- They are the ones who are currently responsible for building internal tools in an enterprise and will continue to do so.
- They understand the development constructs already, and should be successful in building with Budibase
- They will have skills to extend the product and therefore will be able to fill the gaps (with code) where Budibase can't go natively or is missing an extension for.
- They will be the ones to contribute back to our ecosystem of integrations, components etc. - thus growing the overall ecosystem

**IT Professionals**
Examples of this persona may be a network admin, system admin, IT manager, developer in a previous career stage etc. It's "a tech-savvy individual who understands the constructs of development."

This persona should be able to get very far, building with Budibase. However, in our early days, when we still have a small ecosystem of integrations etc, they are likely to hit a wall (or need a developer's help) earlier. E.g. If we don't have a Salesforce integration, and they need one, they need a developer.

So, as time goes on, and our ecosystem grows, the IT professional should accomplish more and more with Budibase.

Thanks, I would love to hear your thoughts!

**Reply by SOVLOOKUP** [reply]
@mjashanks I think he is not only referring to developer and IT Professional. 
Now that Internet is being integrated into every industry, organizations,companies,government and individuals all have great demands for software. I'm a student majoring in law. Students and teachers(not only from Science or Technology department but Education/Agriculture/Management department) in my school need computers to do statistics/data-analysis/display tasks, so they also learn tech like front, database, also use Python, JavaScript and other languages, but due to the limitations of time and energy, they can only make some simple pages and often not up to their own needs. Yes! I was one of them, to make what I want I have spent so much time learning knowledge across full stack😭. But actually budibase can solve 80% needs and save 80% time of mine, I just didn't know it before.

What's more, government in china published the policy of "Internet Plus" in 2015 and is committed to integrating Internet with traditional industries. Today, computer technology is reforming various industries. However, not every company and organization has talents to develop and maintain systems, compared with expensive software procurements, they prefer cheap and efficient cloud services and SaaS, but many custom requirements still can't be met.

Thanks to modern education and technology. Most people around me can understand what is the web, database and front even if they have never studied computer deeply. The code may be still very abstract to them but the visual design(no-code/low-code) is different. As long as a small amount of guide is needed, Users can skip  IT companies' work procedures in the traditional software delivery mode and directly create what they want through the tools built by developers. This may be a better thing for opensource, it benefits both of developers and users.

Some of the Internet business giant in China are developing this kind of business, certainly they Not Open Source. For example:

- Alibaba: https://yida.alibaba-inc.com/
- Baidu: https://suda.baidu.com/

These products are very similar to budibase, their target customer is enterprise, their goal is to change the traditional software delivery mode with SaaS and no-code/low-code design, they are useful but very expensive.

For these users(individuals and enterprise), a low-code solution may be a better choice than software customization because it is cheap, flexible and fast.
For developers and IT Professional, low-code platform is a optional tech to make product.

More people are eager to use budibase.

**Comment by jsmithdev** [comment]
Eager to see what _External data connectors_ comes out as. 

First thought was, since most project is JS, maybe it'll be a JS config file that exports a plugins array...

`external-connectors.config.js`
```js

import { SalesforceConnector } from 'budibase-salesforce-connector';

export const plugins = [ SalesforceConnector ];
```

or traditional Node:
```js
const SalesforceConnector = require( 'budibase-salesforce-connector' );

const plugins = [ SalesforceConnector ];

module.exports = { plugins };
```

In the thought example, `budibase-salesforce-connector` would be a community made npm package.

Not sure the depth of what the object SalesforceConnector would need to contain/expose but i'll check back in. i've used an OSS project called `jsforce` to communicate with Salesforce, metadata, etc plenty so my first try, i'd try to leverage that as a SalesforceConnector POC -- if i'm even remotely close ;) 

either way, cool project :+1:

**Comment by dncpax** [comment]
Hi there. Would love to see possible extension hooks in bb that allow using some libs that are not related to svelte, like openlayers.org.
Ideally I would like to keep away from svelte/react/vue/newkidinblock, that's the reason to use something low-code like bb.
Is there any possibility of having a js component where a few functions would allow for communication back and forth with bb? Like a simple API users could comply with so this would work? Besides these functions, the user would be free to build whatever in that component. So, for example, I could have a map below a table, click on a row button in the table, and zoom the map to that record (this would call a funcion inside the component, following the simple bb signature). Also, the other way around would be to click on a map, and call bb to do something with this value - for instance, change state or emit an event - that would change something in the gui.

**Comment by lollococce** [comment]
Hi!

Annotating here an Idea that I think is the one that has the **highest ROI per difficulty of implementation ratio**

As a quick way to start extending BudiBase, I was thinking of starting to give the possibility of **saving a component created in the BB Builder for later use in other apps**. I know right now one can duplicate or copy paste it within the same app but one cannot save it in a **'Custom' menu next to the built in Svelte components (like Forms, Data, Elements) for later use**. That would let a very fast abstraction and reusability of created components with custom style and should not be difficult to implement.

It also can open a **marketplace of components built on BB Builder** by users and make the development even faster. To really empower the low code philosophy it would also makes sense to me to give the possibility to eventually give the components' creators the possibility to **kind of pack a component to expose the user input needed** for a component only at the highest level of the component tree, so the user of the component can use it with a plug in of values instead of understanding pre-done higher level components or navigating the inner hierarchical complexity of a BB Component, that in the builder quickly becomes messy with higher level of subcomponents and tree depth.

**Reply by mjashanks** [reply]
I think that is a fantastic idea, and especially love the extension to the marketplace.

Reusable components are definitely coming - we've actually done a PoC of it already. Not sure when we will get it fully delivered.

**Reply by gauravpay3ds** [reply]
> Reusable components are definitely coming - we've actually done a PoC of it already. Not sure when we will get it fully delivered.

Is this available (Reusable components are definitely coming )?

**Reply by CallMeLaNN** [reply]
Extension to the marketplace is nice like Strapi. Directus was planned but they're not there yet. One step before that could be about how to share a components. e.g. through zip, github link or even better npm package.

Directus have a simple limited but working extensions, I followed their progress, they don't focus on this much but the existing implementation is enough. I can add a custom page, edit/view/list components, custom REST and hooks for each db CRUD as interceptor callback. They not expose everything to make it simpler but an API (std way) of how they can dynamic import it. See their `app/src/*/register.ts`

I believe Strapi use dynamic import based on folder structure. Their plugins very flexible and can override build-in components. I don't like that because it can be problematic.

TL;DR. Thinking of extension/plugin support could be overwhelming, but dynamic import + svelte props is actually easier and the Reusable Component is one step closer 👍

**Comment by nmstoker** [comment]
Hi @mjashanks - I'm evaluating Budibase for a project and found this discussion because I'm keen to know about how one might extend functionally to cover bespoke needs.

Is there any documentation on how to do this available?

My thinking is that Budibase would be ideal for 90-95 of this particular project (fulfills the needs excellently) but as always happens, it's inevitable we'll need something that's not "off the shelf" in a few screens and I'm keen to understand how feasible this is. Keen to understand how "locked in" one might be and what level of technical challenge it'd be. I looked over the docs but didn't seem to find anything suitable (unless I missed it?)

Thanks in advance. Looks like a great platform you've build here!

**Reply by mjashanks** [reply]
Hi Neil. We will absolutely add the ability to build your own custom UI components and datasources. I'm hoping that we get it done in this half of the year - but it's not officially planned.

**Reply by nmstoker** [reply]
That's great to hear.. Many thanks! I'll be sure to keep an eye out for those changes in due course

**Comment by gzqn** [comment]
I like budibase because it is low code, I am not a coder nor do I want to expend the time to be one. So anything that reduces the need for code is what I want. So a repository of scripts, that have a simple way to be changed to fit my tables and fields and required actions would be useful. I have seen code snippets in the discussion threads needed to make things function properly but finding them again was beyond my skill or time commitment level.

Having components or plugins as mentioned above would also  be useful

**Comment by CallMeLaNN** [comment]
Budibase open source low-code solve a lot of problem in building web apps giving a good promise for my next project.

However I'm quite shock that extensions/plugins is not currently supported yet. My first a few questions that require to build an app is not supported after reading the docs and trying the apps:

1. How to override the CSS, theme and the UI?
  Every apps is unique to their own requirement/branding etc especially about the appearance that all level of people can see. We can't have everything the same. The theme settings currently very limited and basic.

2. How to add my own components?
  Currently only markdown viewer. What if I require markdown editor for simple content management? I will be out of luck. Having extensions/plugins to add external or custom components is necessary before I comfortable to use or recommend this to anyone.

## My suggestions

1. Please reconsider this and discuss the priority or at least make manual extension possible to prevent Budibase user stuck with existing components and UI. Not to mention marketplace or plugin list like suggested by others.

2. Provide integration to n8n just like you have with Zapier.
  What you have in the the CRUD and hook automation is great. Integrating with open source workflow n8n is the best especially the Zapier business model is not friendly to SME. More over n8n offer flexible and limitless automations.

3.  Not related to extension but worth to mention, the app concept is good for some cases not the others.
  Not many headless CMS offer something like this, this is great. But this introduce unnecessary separation for simple system and small team. What if I only have one app? So the portal list of apps in the entry after user login is just an extra navigation step. Otherwise a quick menu to switch between apps is useful. Like switch workspaces in some system or switch servers in Discord. I want to make user less friction to perform their task in the app. Once user login, they can see what they supposed or expect to see.

**Reply by mjashanks** [reply]
Thank you @CallMeLaNN - this is really useful. We really want to get this piece done - it's just a lot of work.

**Comment by marco-santulli** [comment]
Hi Mike,

I think there are various levels of customization potential for BB, I think in order of interest at least for me (difficult to generalize) would be:
1. customization of the CSS, and potential access to HTML code
2. creations of javascript functions inside a BB (with a FaaS model). I think this would fit better the low-code paradigm, because it doesn't need external IDE's. And you already have the basics there in the javascript bindings.
3. improve the debuggability and functionality of existing automations, adding also an external open source workflow (e.g. n8n like someone suggested earlier)
4. development of node.js plugins (I hate js, much prefer java, but can live with that :))
5. development of custom UI widgets (svelte)
7. marketplace

**Reply by mjashanks** [reply]
Hey Marco. I mostly agree - thank you for your feedback. 

1. Customisation of CSS. I'm not sure if we will open up full CSS access, but we will be looking into more granular styling in the next few months. Editing HTML is harder as it's all generated by the components.
2. FaaS. This is a nice idea. The issue here would be around importing of 3rd party libraries - is this something that you would expect to be there? We do have the JS automation block, is there a reason why this is not enough? Also - I'm assuming that you mean this to be backend functions only?
3. Automations debugging. 100% agree - this is a pain right now
4. Yes - we will be adding the ability to add your own custom datasources. Hopefully this year.
5. really hoping to get custom components done this year.
6. Marketplace is a long term goal, but only really relevant once we have custom components and datasources.

Thank you!

**Reply by marco-santulli** [reply]
Hi Mike,

1. I agree, HTML is probably a very long shot and probably HTML widgets would overcome most issues.  I was hoping that CSS would be simpler.
2. I think FaaS would give budi an edge because it would enable coders to work around issues they may find during implementations. I thought something like:
- a REPL/jsfiddle like javascript editor with modules. I have seen there are a couple of open source ones, it may not work out too difficult
- probably as you say server-side but could work also for client side
- for example appsmith has javascript objects you can pass around, something like that would be useful.
- the FaaS should have access to the native javascript objects that you already have (like the data sources or selected rows etc).
M

**Reply by marco-santulli** [reply]
hi, I wanted to add one point on the "extensions".  I think it would be good to have a "custom widget" where one could simply enter the svelte code, without having to create a new widget which requires a lot of "extra" work to make it work with the editor.  My feeling is that 90% of the time one just needs a custom widget in one place so having to go through the extra effort is good if you want to use something really reusable, but a lot of overhead for the "typical" I need just something special here.

**Comment by julrichkieffer** [comment]
While Data / Design / Automation is applicable to every app, the better the developer or more complex the app, the more a 4th aspect becomes conspicuous in its absence: Model.

By Model, I neither mean modelling of data / UI or the Model in MVC. Rather, I mean business modelling as a result of analysis, and making this operational (and central) to an application.

Several examples to illustrate (the absence of an independent Model):

### Options

In the Data layer, a field can be defined as an Option type. A set of accepted values can be specified, and each can be configured with a colour.  However, values != display text, particularly when the user and developer don't share the same language / culture.  Then Budibase's default (and only) approach to displaying the values in the Screen fields highlights the lack of distinction with a display label (commonly provided in other tools). 

Moreover, while an Option field displays all values conifigured, no there's no ability to disallow / prune values at runtime. And, despite configuring a colour per value, either field nor table columns have a means to reflect the associated colour.

### Attachment

An attachment data type has some documented constraints, but conversely to other data types (e.g. Options, Numbers), Attachments rely entirely on the form component to set constraints every time! And even with acceptable file extensions configured, these can be bypassed by drag-n-drop!

So every time the same data field is editable, the extensions and multiplicity must be re-specified.  Whereas DRY would prescribe the configuration be centralised and reused wherever the same data field is affected across screens, forms, or grids.

### Data column names versus form field labels

Similar to Option values, the column names in the data layer become the default label values every single time the field is referenced in screens, forms, tables or grids.  So the UI label must be re-specified, along with placeholder, help text, default value, etc each and every time.

### Data state versus business state

Typically, a record of data includes 1+ state flags.  For example, BudiBase Users have a `State` (Options) field.  But what does the value `active` mean when it's just the value stored in a database?  Enter the (absent business) model.  `active` means x, y but not z actions are allowed by / to that user record. But where is this evident? It's littered throughout the application: in every form / button / automation, the criteria, logic, and mutuation(s) are repeated as needed.

Which drives to another complication:  data nullability versus values becomming required when in particular business states. For example, `If (state=X or Y) and (field=criteria) then fields [a, f, h] are required otherwise can be empty` is common, and the above code both implements **and** assumes this. But what about data imports, API updates, automations or new screens / new developers? When inconsistent -- when logic & mutations are not maintained throughout the app and accross time and interfaces -- then it's quite likely runtime issues and/or data corruption may arise. It's also likely that calculations will be incorrect, eg `SUM(x) / x.Rows Length`.

Thus "spaghetti code" and "a maintenance nightmare" quickly arises, with little to no built-in testability.  Conversely, [a model may be implemented as a state machine](https://xstate.js.org/), both centralising the criteria, logic and mutation(s), and making this DRY, apparent, and testable.... maintainable... and extensible.

So, notwithstanding my implementation suggestion, I advocate more for a Business Model aspect to be independently configured yet integrated into each of the existing Data, Design and Automations layers.
